File: lib/resilint.rb
require 'rest-client'
require 'json'


class Resilint
  def self.registered(opts)
    new(opts)
  end

  attr_accessor :base_url, :user_id, :user_name, :timeout, :post_registration

  def initialize(opts)
    self.base_url          = opts.fetch :base_url
    self.user_name         = opts.fetch :user_name
    self.timeout           = opts.fetch :timeout
    self.post_registration = opts.fetch :post_registration, Proc.new {}
    self.user_id           = opts[:user_id] || register
  end

  def excavate
    json = RestClient::Request.execute(
      method:  :post,
      url:     "#{base_url}/v1/excavate",
      timeout: timeout,
    )
    # => {"bucketId"=>"499728b5-c311-4c59-ac3d-132686dfa036", "gold"=>{"units"=>4}}
    body = JSON.parse(json)
    body.fetch 'bucketId'
  rescue RestClient::Exceptions::ReadTimeout
  end

  def store(bucket_id)
    body = RestClient.post "#{base_url}/v1/store?userId=#{user_id}&bucketId=#{bucket_id}", {}
    body == 'true' or raise(body)
  end

  private

  def register
    json = RestClient.post "#{base_url}/v1/register?userName=#{user_name}", {}
    # => "{\"user\":\"e707b38c-1d53-4be2-a4c7-dd3fc9fc77b8\",\"name\":\"JoshCheek\"}"
    body = JSON.parse(json)
    user_id = body.fetch 'user'
    post_registration.call(user_id)
    user_id
  end
end

File: src/resilint.js
'use strict'

module.exports = Resilint

const https = require('https')

function Resilint(options) {
  const agent = new https.Agent({keepAlive: true})

  const resilint = {
    baseUrl:          options.baseUrl,
    userName:         options.userName,
    userId:           options.userId,
    postRegistration: options.postRegistration,
    https:            https,
    agent:            agent,
    excavate:         excavate,
  }

  if(!resilint.userId) {
    requestUserId(resilint, function(userId) {
      resilint.userId = userId
      resilint.postRegistration(userId)
    })
  }

  return resilint

  function excavate({success, failure, ensure}) {
    const requestOptions = {
      agent:    resilint.agent,
      hostname: resilint.baseUrl,
      method:   'POST',
      path:     '/v1/excavate',
      headers:  {
        'Content-Type':   'application/x-www-form-urlencoded',
        'Content-Length': '0'
      },
    }
    const errorCb = function(err) {
      failure(err)
      ensure()
    }
    const timeoutCb = function() {
      console.log(`timeout`)
    }

    const requestCb = function(res) {
      if(res.statusCode != 200) {
        errorCb(new Error(res.statusMessage))
        return
      }
      let body = ""
      let error = null
      res.setEncoding('utf8')
      res.on('data', (chunk) => body += chunk)
      res.on('error', (err) => {
        error = err
        errorCb(err)
      })
      res.on('end', () => {
        if(error) return
        const parsed   = JSON.parse(body)
        const bucketId = parsed.bucketId
        const type     = "gold" in parsed ? "gold" : "dirt"
        const units    = parsed[type].units
        const value    = units // FIXME: should be zero for dirt
        success(bucketId, type, units, value)
        ensure()
      })
    }

    const req = https.request(requestOptions, requestCb)
    // console.dir(req)
    req.on('error', errorCb)
    // req.setTimeout(1000, timeoutCb)
    req.end()
  }
}


function requestUserId(resilint, cb) {
  const requestOptions = {
    agent:    resilint.agent,
    hostname: resilint.baseUrl,
    method:   'POST',
    path:     `/v1/register?userName=${resilint.userName}`,
    headers:  {
      'Content-Type':   'application/x-www-form-urlencoded',
      'Content-Length': '0'
    },
  }
  const requestCb = function(res) {
    if(res.statusCode != 200) { } // maybe throw or smth?
    let body = ""
    res.setEncoding('utf8')
    res.on('data', (chunk) => body += chunk)
    res.on('end', () => cb(JSON.parse(body).user))
  }

  const errorCb = function(err) {
    console.log(`problem with request: ${err.message}`)
  }
  const timeoutCb = function() {
    console.log(`timeout`)
  }

  const req = https.request(requestOptions, requestCb)
  req.on('error', errorCb)
  // req.setTimeout(1000, timeoutCb)
  req.end()
}
